### PR TITLE
adding support for custom templates

### PR DESCRIPTION
**Pull request by shubhamarora**
## What's changing

*Changes to present POC of impl. More changes are yet to be done.

**Change request: https://github.com/intuit/ReplayZero/issues/6**

Adding support for user to provide their own custom templates. After this change user will be able to provide `path/to/custom/template` and `extension` via parameters when running replay-zero.

Sample command: `replay-zero --output=/path/to/templates/gatling_default.template --extension=scala`

## What else might be impacted?

...

## Checklist

[ ] Unit tests (updated and/or added)
[ ] Documentation (function/class docs, comments, etc.)

### CODE REVIEW

**Comment by Celeo** [comment]
Hi, thanks for the PR! We'll have a review over it soon.

File: main.go [hunk 1]
```diff
@@ -100,25 +102,35 @@ func readFlags() {
 	} else if flags.batchSize > 1 {
 		log.Printf("Using fixed batch size of %d\n", flags.batchSize)
 	}
+
+	// check if template exist at the provided path
+	// TODO: add validation for correctness of template
+	if flags.output != "karate" {
```
Comment (Celeo) [inline main.go:108]: If we are going to support just karate as a built-in, then this should be okay. If we look to support more, though, we'll probably want to hoist those template names into a global variable so we can use it here and in the flag description.

File: templates/gatling_default.template [hunk 2]
```diff
@@ -0,0 +1,66 @@
+package com.yourcompany.perf.yourservice
```
[the remark this answers is not in the document]
Comment (shubhamarora) [reply templates/gatling_default.template:1]: @Celeo My suggestion to this is, that we should stop shipping `.go` files as default supported template and convert them to `.template`. I did this to demonstrate the same. And a user can have more templates like this. Let me know if i make sense.

File: main.go (hunk 1 above)
Comment (shubhamarora) [reply main.go:108]: I keep that open to you and @wtait1 to decide the default supported template. My idea is to go ahead with karate as default, ship it with package and for rest, user can pass path to template.

**Comment by shubhamarora** [comment]
@wtait1 We have some decision awaiting, can you please share you views and I can take this change to fruition. Thanks! :)

File: templates/gatling_default.template (hunk 2 above)
[the remark this answers is not in the document]
Comment (wtait1) [reply templates/gatling_default.template:1]: @shubhamarora yes I think moving to `.template` files is great! In fact, what I'm thinking is that the Karate template should be moved over now as well. Then all test generation is being driven by a `.template` file, but the "default" formats use `.template` files that are shipped with the binary, and custom templates are provided via the new flag.

For packaging the default templates into the binary, there are several tools that do this. `pkger` seems like a good choice (`go-bindata` is no longer supported, and `pkger` is intended as a replacement for `packr`).
https://github.com/markbates/pkger

I've never personally worked with `pkger` before, but I'd be happy to help out / provide guidance as I think creating this common `.template`-based approach would be slick

File: main.go (hunk 1 above)
Comment (wtait1) [reply main.go:108]: I say we should keep both for now - this task is more about changing how templates are provided to the test generation process. See my longer comment below for more detail on how I think we should ship the default templates

**Review by wtait1 — COMMENTED** [review]
Great work so far Shubham. I think there's a bit more discussion we can have to unify the method for providing templates. See detail in the comments

File: main.go [hunk 3]
```diff
@@ -81,10 +82,11 @@ func readFlags() {
 	}
 	flag.BoolVarP(&flags.version, "version", "V", false, "Print version info and exit")
 	flag.IntVarP(&flags.listenPort, "listen-port", "l", 9000, "The port the Replay Zero proxy will listen on")
-	flag.IntVarP(&flags.defaultTargetPort, "target-port", "t", 8080, "The port the Replay Zero proxy will forward to on localhost")
+	flag.IntVarP(&flags.defaultTargetPort, "target-port", "p", 8080, "The port the Replay Zero proxy will forward to on localhost")
```
Comment (shubhamarora) [inline main.go:85]: Need to update doc about this, if this is mentioned somewhere

File: main.go [hunk 4]
```diff
@@ -81,10 +82,11 @@ func readFlags() {
...
 	flag.BoolVar(&flags.debug, "debug", false, "Set logging to also print debug messages")
 	flag.IntVarP(&flags.batchSize, "batch-size", "b", 1, "Buffer events before writing out to a file")
-	flag.StringVarP(&flags.output, "output", "o", "karate", "Either [karate] or [gatling]")
+	flag.StringVarP(&flags.template, "template", "t", "karate", "Either [karate] or [gatling] or [path/to/custom/template]")
+	flag.StringVarP(&flags.extension, "extension", "e", "", "For custom output template")
```
Comment (shubhamarora) [inline main.go:89]: I will update documentation and add details about them after change is finalised

File: templates/gatling_default.template (hunk 2 above)
[the remark this answers is not in the document]
Comment (shubhamarora) [reply templates/gatling_default.template:1]: @shubhamarora done

**Comment by shubhamarora** [comment]
@wtait1 @Celeo I think we should not commit vendor folder. `go.mod` will take care. This will decrease repo side. What do you think?

**Comment by Celeo** [comment]
I believe committing the vendor folder was a decision made when the project was still internal-only. We can revisit it now if there's a notable benefit.

File: main.go (hunk 3 above)
Comment (wtait1) [reply main.go:85]: See latest review comment for more on docs

File: main.go (hunk 4 above)
Comment (wtait1) [reply main.go:89]: See latest review comment for more on docs

**Review by wtait1 — COMMENTED** [review]
@shubhamarora this is shaping up very well, we're almost there!

1. RE: documentation, we do try to make the tool pretty self-describing via good `-h` output 🙂   but we do also have a section in the README on [`different output formats`](https://github.com/intuit/ReplayZero#different-output-formats). If you want to add a sub-section there on passing custom templates, that would be awesome.

2. RE: the `vendor` folder, yes it would be nice to get rid of that but as Matt said, that's an artifact from our previous development process. I'm inclined to say don't worry about it for now.

File: README.md [hunk 5]
```diff
@@ -176,15 +176,23 @@ e46bfc69b893b659c4db713ba2cc84f1 GET /some/other/api
...
 * Gatling (`*.scala`)
 
 Specify the output as a lowercase input to the flag:
 
 ```sh
-replay-zero --output=gatling
+replay-zero --template=gatling
+```
+
+You can also pass path to your own custom template (in case you dont want to use karate or gatling) to the same paramater and `--extension` or `-e` to pass extentsion of the output.
```
Comment (wtait1) [inline README.md:190]: I give this PR a green light after 2 quick additions

1. Can we mention here that the templates are following the format from the [`text/template`](https://golang.org/pkg/text/template/) package in the Go stdlib? So readers have more guidance on where to start
1. While we're at it, let's also let them know that we support all template functions provided by [Sprig](http://masterminds.github.io/sprig/) (ex. "You can even extend your template's functionality with string manipulation, math operators, and more from the Sprig library"), so they don't end up re-inventing like the wheel like I originally did when this project was started!

Comment (shubhamarora) [reply README.md:192]: Done!

**Review by wtait1 — APPROVED** [review]
LGTM. Loved the collaboration / conversation in this PR to bring a great extension to Replay Zero's functionality!

**Review by Celeo — APPROVED** [review]
Fantastic work! We really appreciate your contribution!

**Comment by wtait1** [comment]
:rocket: PR was released in `v0.1.0` :rocket:

**Pull request merged**